### PR TITLE
[processing] Improve temporary directory handling

### PR DESCRIPTION
**Pull request by nyalldawson**
- Make the new Temp Folder setting optional, default to an empty string, and note that empty = use default
- Move responsibility for cleanup of temporary folders to c++ (if Processing Python part crashes, or doesn't exist, then we still want these cleaned up correctly)

### CODE REVIEW

**Comment by nyalldawson** [comment]
@alexbruy what do you think?

**Review by alexbruy — COMMENTED** [review]
Looks good. I was about to look into moving cleanup steps to core code too, but you are faster :-).
+1 to merge

**Review by alexbruy — APPROVED** [review]

**Comment by nyalldawson** [comment]
@alexbruy @elpaso @nirvn @m-kuhn do you mind re-reviewing this update? This code makes me very nervous, as I'm concerned that a bad setting as the temp folder could potentially result in data being wiped (e.g. if a user set the path as their documents folder). This change **should** be safe, because we append the /processing_XXXXXX subpath to whatever the user sets, but I'd really value extra sets of eyes over the logic to make sure I haven't missed anything.

**Comment by alexbruy** [comment]
Looks good to me, as far as I can see it should be safe.

File: src/core/processing/qgsprocessingutils.cpp [hunk 1]
```diff
@@ -746,17 +746,34 @@ QVariant QgsProcessingUtils::generateIteratingDestination( const QVariant &input
...
+  }
+  else if ( sFolder.isEmpty() || !sFolder.startsWith( basePath ) || !sTempFolder )
+  {
+    if ( !QDir().exists( basePath ) )
+      QDir().mkpath( basePath );
+
+    const QString templatePath = QStringLiteral( "%1/processing_XXXXXX" ).arg( basePath );
+    // leak the previous folder -- we don't want it to be cleaned up, we don't know what was in it that may still
+    // be required for this session!
+    sTempFolder.release();
```
Comment (m-kuhn) [inline src/core/processing/qgsprocessingutils.cpp]:
```suggestion
    sTempFolder.setAutoRemove( false );
```

If I read this correctly, this currently leaks memory next to a physical folder.

Comment (m-kuhn) [reply src/core/processing/qgsprocessingutils.cpp]: Alternatively, keep a static list of previous temp folders somewhere to be destroyed on application exit.

Comment (nyalldawson) [reply src/core/processing/qgsprocessingutils.cpp]: Good idea! Can you check the new version?

**Pull request merged**